dso-l1/src/main/java/com/tc/object/ClientEntityManagerImpl.java

```diff
@@ -574,27 +584,32 @@ private NetworkVoltronEntityMessage createMessageWithoutClientInstance(EntityID
   }
 
   private NetworkVoltronEntityMessage createMessageWithDescriptor(EntityID entityID, EntityDescriptor entityDescriptor, boolean requiresReplication, byte[] config, VoltronEntityMessage.Type type, Set<VoltronEntityMessage.Acks> acks) {
-    // Get the clientID for our channel.
+    NetworkVoltronEntityMessage message = (NetworkVoltronEntityMessage) channel.createMessage(TCMessageType.VOLTRON_ENTITY_MESSAGE);
     ClientID clientID = this.channel.getClientID();
-    // Get the next transaction ID.
-    TransactionID transactionID = new TransactionID(currentTransactionID.incrementAndGet());
-    // Figure out the "trailing edge" of the current progress through the transaction stream.
-    TransactionID oldestTransactionPending = transactionID;
-    // if reconnectable, discover the oldest transaction still being waited for
-    if (this.channel.getProductID().isReconnectEnabled()) {
-      for (TransactionID pendingID : this.inFlightMessages.keySet()) {
-        if (oldestTransactionPending.compareTo(pendingID) > 0) {
-          // pendingID is earlier than oldestTransactionPending.
-          oldestTransactionPending = pendingID;
+    TransactionID transactionID = transactionSource.create();
+    TransactionID oldestTransactionPending = transactionSource.oldest();
+    //this is expensive -- only assert when testing
+    TransactionID oldestTransactionInFlight;
+    assert (oldestTransactionInFlight = oldestTransactionIn(inFlightMessages.keySet())) == null || (oldestTransactionPending.compareTo(oldestTransactionInFlight) <= 0);
```

Comment:
are asserts off by default?

Reply:
Yes.... build-system test runners tend to turn them on. I was torn between this and eliminating this check entirely. Oldest transaction lookup is now O(1), rather than O(N) where N is the number of inflight txns.